conformance_tests/tools/debug/src/test_debug.cpp

```diff
@@ -1376,6 +1377,143 @@ void zetDebugReadWriteRegistersTest::run_read_write_registers_test(
+
+    ze_device_thread_t device_threads = {};
+    device_threads.slice = UINT32_MAX;
+    device_threads.subslice = UINT32_MAX;
+    device_threads.eu = UINT32_MAX;
+    device_threads.thread = UINT32_MAX;
+
+    LOG_INFO << "[Debugger] Stopping all device threads";
+    // give time to app to launch the kernel
+    std::this_thread::sleep_for(std::chrono::seconds(6));
+    lzt::debug_interrupt(debugSession, device_threads);
+
+    std::vector<ze_device_thread_t> stopped_threads;
+    if (!find_stopped_threads(debugSession, device, device_threads, true,
+                              stopped_threads)) {
+      delete[] kernel_buffer;
+      FAIL() << "[Debugger] Did not find stopped threads";
+    }
+
+    bool is_heapless_mode = false;
+    LOG_INFO << "[Debugger] Reading/Writing Thread Scratch Register on "
+                "interrupted threads";
+
+    for (auto &stopped_thread : stopped_threads) {
+      uint8_t *mode_values = nullptr;
+      auto register_set_properties = lzt::get_register_set_properties(device);
+      for (auto &register_set : register_set_properties){
+
+        if (register_set.type == ZET_DEBUG_REGSET_TYPE_MODE_FLAGS_INTEL_GPU){
+          auto reg_size_in_bytes = register_set.count * register_set.byteSize;
+          mode_values = new uint8_t[reg_size_in_bytes];
+          ASSERT_EQ(
+              zetDebugReadRegisters(debugSession, stopped_thread,
+                                    ZET_DEBUG_REGSET_TYPE_MODE_FLAGS_INTEL_GPU,
+                                    0, register_set.count, mode_values),
+              ZE_RESULT_SUCCESS);
+
+          uint32_t *uint32_t_values = (uint32_t *)mode_values;
+          LOG_DEBUG << "[Debugger] mode value: %u " << uint32_t_values[0];
+          is_heapless_mode =
+              (uint32_t_values[0] & ZET_DEBUG_MODE_FLAG_HEAPLESS);
+        }
+      }
+
+      if (is_heapless_mode) {
+        for (auto &register_set : register_set_properties){
+          if ((register_set.type ==
+               ZET_DEBUG_REGSET_TYPE_THREAD_SCRATCH_INTEL_GPU) &&
+              (register_set.generalFlags & ZET_DEBUG_REGSET_FLAG_READABLE)) {
+            LOG_DEBUG << "[Debugger] Register set type " << register_set.type
+                      << " is readable";
+            size_t reg_size_in_bytes =
+                register_set.count * register_set.byteSize;
+
+            uint64_t *thread_scratch_reg_values =
+                new uint64_t[reg_size_in_bytes];
+            ASSERT_EQ(zetDebugReadRegisters(
+                          debugSession, stopped_thread,
+                          ZET_DEBUG_REGSET_TYPE_DEBUG_SCRATCH_INTEL_GPU, 0,
+                          register_set.count, thread_scratch_reg_values),
+                      ZE_RESULT_SUCCESS);
+          } else {
+            LOG_INFO << "[Debugger] Register set type " << register_set.type
+                     << " is NOT readable";
+          }
+          if (register_set.generalFlags & ZET_DEBUG_REGSET_FLAG_WRITEABLE) {
```

Comment:
Please fix this if else. We could also fail in case we find the reigster as writable

Reply:
Ok